File: WalletTracker/app/src/main/java/com/chtima/wallettracker/dao/CategoryDao.java
package com.chtima.wallettracker.dao;

import androidx.room.*;

import com.chtima.wallettracker.models.Category;
import com.chtima.wallettracker.models.CategoryWithTransactions;

import java.util.List;

import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Completable;

@Dao
public interface CategoryDao {

    @Insert
    Maybe<Long> insert (Category category);

    @Insert()
    Completable insertAll(Category...categories);

    @Query("SELECT * FROM categories")
    Flowable<List<Category>> getAll();

    @Query("SELECT * FROM categories")
    Flowable<List<CategoryWithTransactions>> getCategoriesWithTransactions();

    @Transaction
    @Query("SELECT * FROM categories WHERE id IN (SELECT categoryId FROM transactions WHERE userId = :userId)")
    Flowable<List<CategoryWithTransactions>> getCategoriesWithTransactionsByUserId(long userId);

}

File: WalletTracker/app/src/main/java/com/chtima/wallettracker/db/dao/CategoryDao.kt
package com.chtima.wallettracker.db.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Transaction
import com.chtima.wallettracker.models.Category
import com.chtima.wallettracker.models.Category.CategoryType;
import com.chtima.wallettracker.models.CategoryWithTransactions
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.core.Maybe


@Dao
interface CategoryDao {

    @Insert
    fun insert(category: Category): Maybe<Long>

    @Insert
    fun insertAll(vararg categories: Category): Completable

    @Query("SELECT * FROM categories")
    fun getAll(): Flowable<List<Category>>

    @Query("SELECT * FROM categories WHERE categoryType = :type")
    fun getByType(type: CategoryType?): Flowable<List<Category>>

    @Query("SELECT * FROM categories")
    fun getCategoriesWithTransactions(): Flowable<List<CategoryWithTransactions>>

    @Transaction
    @Query("SELECT * FROM categories WHERE id IN (SELECT `categoryId` FROM `transaction` WHERE userId = :userId)")
    fun getCategoriesWithTransactionsByUserId(userId: Long): Flowable<List<CategoryWithTransactions>>

}